airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py

```diff
@@ -131,40 +132,61 @@ def stream_slices(self) -> Iterable[StreamSlice]:
                 parent_field = parent_stream_config.parent_key.eval(self.config)  # type: ignore # parent_key is always casted to an interpolated string
                 partition_field = parent_stream_config.partition_field.eval(self.config)  # type: ignore # partition_field is always casted to an interpolated string
                 incremental_dependency = parent_stream_config.incremental_dependency
-                for parent_stream_slice in parent_stream.stream_slices(
-                    sync_mode=SyncMode.full_refresh, cursor_field=None, stream_state=None
-                ):
-                    parent_partition = parent_stream_slice.partition if parent_stream_slice else {}
-
-                    # we need to read all records for slice to update the parent stream cursor
-                    stream_slices_for_parent = []
-
-                    # only stream_slice param is used in the declarative stream, stream state is set in PerPartitionCursor set_initial_state
-                    for parent_record in parent_stream.read_records(
-                        sync_mode=SyncMode.full_refresh, cursor_field=None, stream_slice=parent_stream_slice, stream_state=None
-                    ):
-                        # Skip non-records (eg AirbyteLogMessage)
-                        if isinstance(parent_record, AirbyteMessage):
-                            if parent_record.type == Type.RECORD:
-                                parent_record = parent_record.record.data
-                            else:
-                                continue
-                        elif isinstance(parent_record, Record):
-                            parent_record = parent_record.data
-                        try:
-                            partition_value = dpath.get(parent_record, parent_field)
-                        except KeyError:
-                            pass
+
+                stream_slices_for_parent = []
+                previous_associated_slice = None
+                for parent_record in parent_stream.read_stateless():
+                    parent_partition = None
+                    parent_associated_slice = None
+                    # Skip non-records (eg AirbyteLogMessage)
+                    if isinstance(parent_record, AirbyteMessage):
```

Comment:
I think there's a gotcha here. My understanding is we can't map the parent record to the parent slice if it was emitted as a `AirbyteMessage`. I suspect this is not a huge deal since the retriever yields `Record`s but it feels like something worth noting.

Reply:
I would probably at least log a warning with the impact of doing such thing. My understand is that:
* In terms of risk, it can only happen when having a CustomRetriever which I don't see in the airbyte-integration right now so the risk seems almost none
* The impact is that it'll not update the state at every slice which means that if the sync fail, we need to start back from the beginning

Reply:
fair point i'll log a warning, can't hurt. I just sort of copied the old logic over, but doesn't mean it can't be improved. will fix!